### PR TITLE
Fix some bugs in loading OpenGL/GLX/EGL libraries

### PR DESCRIPTION
**Pull request by ya-isakov**
Without additional check, even if libOpenGL was loaded, libGL.so will be loaded as well, and used both in gl_handle and glx_handle, so libglvnd libraries will not be used.

The problem is that on Wayland-only system, there will be no libGL, so epoxy_load_gl will fail even if libOpenGL was loaded.

### CODE REVIEW

**Comment by ya-isakov** [comment]
Hmm, it looks like this leads to not loading GLX library, as epoxy_conservative_glx_dlsym is using `exit_if_fails` as a `load` flag.

**Comment by ebassi** [comment]
I guess this is for @nwnk to review.

**Comment by ya-isakov** [comment]
I've found a way to fix problem with loading libGL/libGLX on every `epoxy_current_context_is_glx` check - there is no point of trying `epoxy_conservative_egl_dlsym` if appropriate handle is NULL.
If this handle is NULL, it means that library was not loaded, and `epoxy_conservative_egl_dlsym` with load=exit_if_fails=false will not load it as well, and this will lead to `do_dlsym` also returning NULL. So, I've added checks for api.*_handle in `epoxy_current_context_is_glx` function.

**Comment by ya-isakov** [comment]
Here is output of strace for https://github.com/ebassi/glarea-example on libglvnd-enabled wayland (libGL.so/libGLX.so still exists):
Before my fixes:
....
[pid 3520958] openat(AT_FDCWD, "/usr/lib64/libOpenGL.so.0", O_RDONLY|O_CLOEXEC) = 22
[pid 3520958] openat(AT_FDCWD, "/usr/lib64/libGL.so.1", O_RDONLY|O_CLOEXEC) = 22
[pid 3520958] openat(AT_FDCWD, "/usr/lib64/libGLX.so.0", O_RDONLY|O_CLOEXEC) = 22

After:
...
[pid 3519200] openat(AT_FDCWD, "/usr/lib64/libOpenGL.so.0", O_RDONLY|O_CLOEXEC) = 22

**Comment by nwnk** [comment]
One thing we might want to consider is using `dlopen(RTLD_NOLOAD)` to probe for things...

**Comment by nwnk** [comment]
Patch 1 2 and 4 in this series are:

`Reviewed-by: Adam Jackson <ajax@redhat.com>`

Patch 3 ("Do not load symbols in epoxy_current_context_is_glx [...]") isn't great. With that change, `epoxy_current_context_is_glx` will fail if it's the very first epoxy library call, even if there _is_ a current context and it _is_ GLX. We should instead gently (RTLD_NOLOAD-style) test whether GLX libraries are already loaded, and fill in `api.glx_handle` and friends based on that.

**Comment by ya-isakov** [comment]
@nwnk I'm completely fine, if you would merge only 3 commits. Third one was my best try on fixing the load of libGL.so on every symbol lookup, if EGL is used, but I'm not sure how to fix it properly.

**Comment by ya-isakov** [comment]
Actually, if epoxy_current_context_is_glx if first epoxy library call, it should not fail, just return false. Oh, and without my commit, it will also return false, as it will use `epoxy_conservative_glx_dlsym("glXGetCurrentContext", false)` -> `epoxy_load_glx(false, false)`, and epoxy_load_glx with load=false is not setting api.glx_handle (if library was not loaded before), so it's result will be false, so `epoxy_conservative_glx_dlsym` will return NULL.

TL;DR - without my commit 3, if `epoxy_current_context_is_glx` is the first function to run, it will return false, the same as with my commit.

**Comment by ya-isakov** [comment]
Oh, and how it could be possible, to have GLX context without loading libGLX first? And to load it, application will use `glXCreateNewContext`, which will use `epoxy_glx_dlsym`, which will load libGLX.so, if needed.

**Comment by nwnk** [comment]
> And to load it, application will use glXCreateNewContext, a wrapper for epoxy_glx_dlsym, which will load libGLX.so, if needed, and set api.glx_handle.

This isn't necessarily true. Imagine plugin library A gets loaded first, which links against libGL directly, creates a GLX context, and makes it current. Then plugin library B is loaded, which links against epoxy.

**Comment by tomeuv** [comment]
@nwnk Could we get 1, 2 and 4 merged? With them, I'm able to run virglrenderer without linking to GLX.

**Comment by ebassi** [comment]
Patch 2 introduces a test failure in `glx_alias_prefer_same_name`.

I've opened #238 with patches 1 and 4.

**Comment by batesj** [comment]
Any thoughts on whether https://github.com/anholt/libepoxy/issues/240 is a libepoxy bug or apitrace?

**Comment by ya-isakov** [comment]
@batesj I would say, that it is probably a apitrace not supporting libOpenGL.so. I cannot find any mentions of it in it's code, so, maybe it's trying to override libGL.so, but with my patch, it should not be used, if libOpenGL.so is present in the system

**Comment by ya-isakov** [comment]
@ebassi I'm fine if commit 2 (libGLX.so version fix) is omitted, it's probably different in different distributions. On GLX-free system there is no such file at all...
@nwnk Sorry, proper fixing of patch 3 is way above my skill set in C :(

**Comment by batesj** [comment]
Can there be a new release to pick up this bug fix?

**Comment by ebassi** [comment]
@batesj I just released 1.5.6 which contains the first and last commits of this MR.

**Comment by ebassi** [comment]
If nobody beats me to it, I'll try to carve some time to rework the third commit to follow the approach outlined by @nwnk.

**Comment by batesj** [comment]
Any thoughts on bringing these fixes back without breaking some distros? Compile option perhaps?
